File: ProblemSolving/450DSA/Python/src/string/countPalindromicSubsequences.py
"""
@author Anirudh Sharma

Given a string s, find number of palindromic subsequence (need not necessarily be distinct)
which could be formed from the string.
"""


def expandFromMiddle(s, left, right):
    count = 0
    # Check for equality
    while left >= 0 and right < len(s) and s[left] == s[right]:
        # Update the count
        count += 1
        # Expand both sides
        left -= 1
        right += 1
    return count


def countPalindromicSubsequences(s):
    # Count of all the palindromic subsequences
    count = 0
    # Loop through the string
    for i in range(len(s)):
        # For odd length string, expand from s[i]
        count += expandFromMiddle(s, i, i)
        # For odd length string, expand from s[i] and s[1+1]
        count += expandFromMiddle(s, i, i + 1)
    return count


if __name__ == "__main__":
    print(countPalindromicSubsequences("abcd"))
    print(countPalindromicSubsequences("aab"))
    print(countPalindromicSubsequences("aaaaaaaaaaaaaaaaaaaaaaaaaaa"))

File: ProblemSolving/450DSA/JavaScript/src/string/CountPalindromicSubsequences.js
/**
 * @author Anirudh Sharma
 * 
 * Given a string s, find number of palindromic subsequence (need not necessarily be distinct)
 * which could be formed from the string.
 */
const countPalindromicSubsequences = (s) => {
    // Count of all the palindromic subsequences
    let count = 0;
    // Loop through the string
    for (let i = 0; i < s.length; i++) {
        // For odd length string, expand from s[i]
        count += expandFromMiddle(s, i, i);
        // For even length string, expand from s[i]
        // and s[i+1]
        count += expandFromMiddle(s, i, i + 1);
    }
    return count;
};

const expandFromMiddle = (s, left, right) => {
    let count = 0;
    // Check for the equality
    while (left >= 0 && right < s.length && s.charAt(left) === s.charAt(right)) {
        // Increment the count
        count++;
        // Expand from both sides
        left--;
        right++;
    }
    return count;
};

const main = () => {
    console.log(countPalindromicSubsequences("abcd"));
    console.log(countPalindromicSubsequences("aab"));
    console.log(countPalindromicSubsequences("aaaaaaaaaaaaaaaaaaaaaaaaaaa"));
};

main();